notebooks/adhoc/jovian_analysis/jovian_src/core.py

```diff
@@ -0,0 +1,1121 @@
+            if show_progress:
+                print(f"❌ Error fetching from GCS: {e}")
+                print(f"💡 Expected directory: {block_data_dir.absolute()}")
+            return pl.DataFrame()
+
+
+# =============================================================================
+# CALLDATA ANALYSIS
+# =============================================================================
+
+@dataclass
+class TransactionAnalysis:
+    """Result of analyzing a single transaction."""
+    transaction_hash: str
+    calldata_size: int
+    fastlz_size: int
+    size_estimate: float
+    calldata_footprint: float
+    compression_ratio: float
+
+
+@dataclass
+class BlockAnalysis:
+    """Result of analyzing a block."""
+    block_number: int
+    tx_count: int
+    total_calldata_footprint: float
+    total_size_estimate: float
+    calldata_utilization: float
+    exceeds_limit: bool
+    avg_footprint_per_tx: float
+    max_tx_footprint: float
+    transactions: List[TransactionAnalysis]
+    footprint_cost: int
+    block_gas_limit: int
+    block_gas_used: Optional[int] = None
+    utilization_vs_gas_used: Optional[float] = None
+    total_calldata_size: int = 0  # Sum of raw calldata sizes
+    total_fastlz_size: int = 0    # Sum of compressed sizes
+
+class CalldataAnalyzer:
+    """Core analyzer for Jovian calldata footprints."""
+
+    def __init__(self, jovian_config: JovianConfig, analysis_config: AnalysisConfig):
+        self.jovian_config = jovian_config
+        self.analysis_config = analysis_config
+        self.compressor = FastLZCompressor()
+
+    def calculate_size_estimate(self, calldata: bytes) -> float:
+        """Calculate Jovian size estimate for calldata."""
+        if not calldata:
+            return float(self.jovian_config.min_transaction_size)
+
+        fastlz_size = self.compressor.compress_len(calldata)
+        size_estimate = max(
+            self.jovian_config.min_transaction_size,
+            self.jovian_config.intercept + self.jovian_config.fastlz_coef * fastlz_size / 1e6
+        )
+        return size_estimate
+
+    def calculate_footprint(self, calldata: bytes, footprint_cost: int) -> float:
+        """Calculate calldata footprint for a transaction."""
+        size_estimate = self.calculate_size_estimate(calldata)
+        return size_estimate * footprint_cost
+
+    def analyze_transaction(self, tx_row: Dict[str, Any], footprint_cost: int) -> TransactionAnalysis:
+        """Analyze a single transaction."""
+        calldata_hex = tx_row.get('input', '0x')
+        calldata = parse_calldata(calldata_hex)
+
+        # OPTIMIZATION: Calculate FastLZ compression once and reuse for all calculations
+        # Original code had 3 redundant computations:
+        # 1. Direct compress_len() call
+        # 2. compress_len() inside calculate_size_estimate()
+        # 3. compress_len() inside calculate_footprint() -> calculate_size_estimate()
+        fastlz_size = self.compressor.compress_len(calldata)
```

Comment:
same here
```suggestion
        fastlz_size = self.compressor.compress_len(calldata) + 68
```
wondering why the logic and calculation effort seem to be duplicated between `analyze_transaction` and `calculate_size_estimate`. The FastLZ compression is not super expensive but also not cheap to perform, so it would be better to deduplicate, also to make sure we have to define the size_estimate calculation only in one place.